### PR TITLE
Introduce and utilise treelib

### PR DESCRIPTION
**Pull request by diivm**
Linking #21

### CODE REVIEW

**Review by kunaltyagi — CHANGES_REQUESTED** [review]
I feel like I've seen this code before. Could you dedup it (and target the PR branch properly if that's the issue_?

**Comment by diivm** [comment]
> I feel like I've seen this code before. Could you dedup it (and target the PR branch properly if that's the issue_?

Branch created from the branch used at #6, needs merge

File: clang_bind/parse.py [hunk 1]
```diff
@@ -1,134 +1,68 @@
...
-            "access_specifier"
-        ].name
-        # replace `TypeKind.value` with just `value`
-        parsed_node["type"]["kind"] = parsed_node["type"]["kind"].name
-
-        return parsed_node
-
-    @classmethod
-    def parse_node_recursive(cls, node):
+    def _is_valid_child(self, clang_child_cursor):
```
Comment (kunaltyagi) [inline clang_bind/parse.py]:
```py
def is_node_from_file(node, filename):
  pass

def _is_valid_child(self, child_node):
  return is_node_from_file(child_node, self.filename)

File: clang_bind/parse.py [hunk 2]
```diff
@@ -1,134 +1,68 @@
...
-        for child in cursor.get_children():
-            child_node = {"cursor": child, "filename": filename, "depth": depth + 1}
-            if cls._is_valid_child(node, child_node):
-                child_parsed_info = cls.parse_node_recursive(child_node)
-                parsed_info["members"].append(child_parsed_info)
-
-        return parsed_info
-
-    def get_parsed_info(self):
+    def _construct_tree(self, node):
```
Comment (kunaltyagi) [inline clang_bind/parse.py:58]: Same here? Make construct tree a free function and get_tree provides an additional argument.

Reason: functionality is independent of class. Extracting out allows a smaller to parse class

Comment (diivm) [reply clang_bind/parse.py:58]: Even if we pass the arguments, we need to call another method either via `self` or `cls`:
https://github.com/PointCloudLibrary/clang-bind/blob/df6cd5809cae7d66b0f2e7538aba6fcc0bddd81d/clang_bind/parse.py#L60

**Comment by diivm** [comment]
@kunaltyagi 

The changes needed in the `generate.py` and `test_generate.py` will be a lot, will address them in another PR.

File: .github/workflows/pytest.yml [hunk 3]
```diff
@@ -27,8 +27,8 @@ jobs:
       - name: Install pip
         run: python -m pip install --upgrade pip
 
-      - name: Install pytest
-        run: pip install pytest
+      - name: Install packages
+        run: pip install pytest treelib
```
Comment (kunaltyagi) [inline .github/workflows/pytest.yml:31]: I think we should add a poetry setup instead. (separate issue)

File: tests/test_parse.py [hunk 4]
```diff
@@ -1,46 +1,27 @@
...
 
-        compilation_database_path = get_compilation_database_path(
-            tmp_path=tmp_path, filepath=source_path
-        )
-
-        compilation_database = CompilationDatabase(
-            compilation_database_path=compilation_database_path
-        )
+    tree = Parse(source_path, ARGS).get_tree()
+    tree_paths = tree.paths_to_leaves()
```
Comment (kunaltyagi) [inline tests/test_parse.py]: Please add some comments on how the tree_paths and tree are related to the AST

File: clang_bind/parse.py [hunk 5]
```diff
@@ -1,134 +1,76 @@
...
 
-import clang_bind.utils as utils
 from clang_bind.clang_utils import ClangUtils
 
 
+class Node:
+    def __init__(self, cursor, verbose=False):
+        self.cursor = cursor
+        if verbose:
+            # checks available in cindex.py via clang_utils.py
```
Comment (kunaltyagi) [inline clang_bind/parse.py]: reword this a little?

File: .github/workflows/pytest.yml (hunk 3 above)
Comment (diivm) [reply .github/workflows/pytest.yml:31]: Created an issue.

**Pull request merged**